### PR TITLE
Add patch method for container images

### PR DESCRIPTION
**Pull request by johscheuer**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR makes it possible to run e2e tests in air gapped scenarios e.g. when images can't be pulled from the internet. One assumption is that newer e2e tests should use the create function in order to create objects from manifest files. I introduced a little helper function that replaces the hardcoded image registry with a custom defined one (passed as environment variable).

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
The k8s Bot just closed my PR so I needed to open this one: https://github.com/kubernetes/kubernetes/pull/74273

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by johscheuer** [comment]
Ping @pohly

File: test/utils/image/manifest.go
Comment (pohly) [inline test/utils/image/manifest.go]: Please use k8s.io/klog as in the rest of Kubernetes (if there is logging instead of error reporting, see other comment).

File: test/utils/image/manifest.go
Comment (pohly) [inline test/utils/image/manifest.go]: This will now show up in all logs (whether `KUBE_TEST_REPO_LIST` is used or not), right? That's probably okay, given that we want developers who run in a non-air-gapped cluster to write their test so that it can also run in an air-gapped cluster.

But given that, should this really just be a log message (which probably will get overlooked) and not an error?

Either way, it's not going to be obvious to developers who see this message a) where it comes from and b) what they need to do about it.

Can you add more details and instructions?

**Review by pohly — CHANGES_REQUESTED** [review]

File: test/utils/image/manifest.go
Comment (johscheuer) [reply test/utils/image/manifest.go]: Hm, I think I would favour a more descriptive error. I agree with you that probably nobody reads the log message if tests are passing.

Comment (pohly) [reply test/utils/image/manifest.go]: Sounds good to me.

Comment (johscheuer) [reply test/utils/image/manifest.go]: I updated the code, PTAL.

File: test/e2e/framework/create.go
Comment (pohly) [inline test/e2e/framework/create.go]: Why?

It's generally a good idea to explain the "why" in the commit message or in the code, because then reviewers don't need to ask ;-}

Hmm, it looks like this commit fixes the previous one. I think you can just squash the two.

File: test/utils/image/manifest_test.go
Comment (pohly) [inline test/utils/image/manifest_test.go]: Is this TODO still valid?

File: test/utils/image/manifest.go
Comment (pohly) [inline test/utils/image/manifest.go]: Let's add "add the registry, *otherwise the test will fail on air-gapped clusters*". Developers are likely to be more motivated to make that change when they know that there is a reason for it.

**Review by pohly — CHANGES_REQUESTED** [review]
The commit message for f9b0f9ac7db773b972cc32a5bd6c9989ee0b48d6 could be better.

There's a typo ("is" instead of "if") and no explanation why this is a useful change. You and I know it because we discussed it in the PR, but anyone else looking at this code will not.

File: test/utils/image/manifest_test.go
Comment (johscheuer) [reply test/utils/image/manifest_test.go]: No, removed it. Good catch!

File: test/utils/image/manifest.go
Comment (johscheuer) [reply test/utils/image/manifest.go]: Addeed

**Review by pohly — APPROVED** [review]
You might have a look at recommendations for writing good commit message (for example, https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53). But Kubernetes doesn't have official commit message guidelines, so...

\lgtm

**Comment by johscheuer** [comment]
Thanks for your link ! I will try to write better commit messages in the future.

File: test/e2e/framework/create.go
Comment (pohly) [inline test/e2e/framework/create.go]: Why the package rename? The author of that package called it just 'image'. Unless there's a good reason (like a conflict with some other package) I wouldn't change that.

Comment (johscheuer) [reply test/e2e/framework/create.go]: I just oriented at the other packages consuming utils/image, all of these use this package rename (a `grep -R -i -I --include "*.go" "\/test\/utils\/image" test` shows this. I thought this was some kind of convention. I'm fine with changing this.

Comment (pohly) [reply test/e2e/framework/create.go]: No, if everyone does that, then keep it.

**Comment by johscheuer** [comment]
I updated the code and ran `make verify` which seems to be happy now there are only some errors in the following scripts in totally unrelated code parts:

```bash
hack/make-rules/../../hack/verify-golint.sh
hack/make-rules/../../hack/verify-govet.sh
hack/make-rules/../../hack/verify-openapi-spec.sh
hack/make-rules/../../hack/verify-typecheck.sh
```

**Comment by johscheuer** [comment]
@pohly any further feedback on this PR ? Otherwise a `/ok-to-test` would be nice so I can see that all test also pass in prow.

**Comment by johscheuer** [comment]
Since 1.14.0 is out can we merge this ? This allows us to run the conformance tests again in our air-gaped clusters.

**Comment by johscheuer** [comment]
ping @pohly :)

**Comment by pohly** [comment]
> ping @pohly :)

I wasn't allowed to approve at the time of the initial review, but now it should work (https://github.com/kubernetes/kubernetes/commit/e0f0d4c9f8496f6645340321df4e2f353b0f6f73#diff-8a7b63a12fc7b35786fb54454bd3d7da), so:

/approve

**Comment by johscheuer** [comment]
@pohly it seems like your approve command didn't worked / wasn't recognized by the bot

**Comment by pohly** [comment]
This touches files under `test/utils`. I can't approve those changes.

**Comment by johscheuer** [comment]
ping @luxas @mkumatag @ixdy @listx (I took you from the approvers file)

Could someone take a look ?

**Comment by mkumatag** [comment]
> ping @luxas @mkumatag @ixdy @listx (I took you from the approvers file)
> 
> Could someone take a look ?

@johscheuer This option is already provided via `KUBE_TEST_REPO_LIST` part of https://github.com/kubernetes/kubernetes/pull/60848 and can refer https://github.com/heptio/sonobuoy/issues/160#issuecomment-434215877 to understand how to use it.

hence closing this PR.

/close

**Comment by pohly** [comment]
@mkumatag I suspect that you closed this PR a bit too quickly. The code that @johscheuer is enhancing here currently ignores this `KUBE_TEST_REPO_LIST` and directly uses the registries specified in the .yaml files.

/reopen

**Comment by pohly** [comment]
@johscheuer perhaps update the PR description to clarify *which* E2E test now work in air-gapped clusters thanks to your PR. Re-reading the description I can see how @mkumatag arrived at the conclusion that this is a duplicate of already existing functionality: "this PR makes it possible to run e2e tests in air gapped scenarios" - clearly that is already possible, just not for all of them.

**Comment by mkumatag** [comment]
> @mkumatag I suspect that you closed this PR a bit too quickly. The code that @johscheuer is enhancing here currently ignores this `KUBE_TEST_REPO_LIST` and directly uses the registries specified in the .yaml files.

In my opinion, we need to start converting those `.yaml` into templates and replace the image repositories via manifest, e.g: https://github.com/kubernetes/kubernetes/blob/master/test/fixtures/doc-yaml/user-guide/liveness/http-liveness.yaml.in#L11

**Comment by pohly** [comment]
> In my opinion, we need to start converting those .yaml into templates and replace the image repositories via manifest

Those .yaml files are verbatim copies from some other repo which doesn't use templating. It would be awkward to test something different in Kubernetes. I prefer the approach taken in this PR.

**Comment by mkumatag** [comment]
> > In my opinion, we need to start converting those .yaml into templates and replace the image repositories via manifest
> 
> Those .yaml files are verbatim copies from some other repo which doesn't use templating. It would be awkward to test something different in Kubernetes. I prefer the approach taken in this PR.

okay, make sense. I'm okay with the changes. Any other opinion @ixdy?

**Comment by johscheuer** [comment]
@pohly do you think it would be a good idea to create an issue to track the conversion of the yaml files into the templating format ? I would start working on this (as soon as this PR is merged).

**Comment by pohly** [comment]
Johannes Scheuermann <notifications@github.com> writes:
> @pohly do you think it would be a good idea to create an issue to
> track the conversion of the yaml files into the templating format ?

I would prefer to keep the yaml files as they are, because then it is
easier to keep them synchronised with the corresponding files in the
csi-driver-host-path repo. If we were to make changes related to them,
then it should be around automatically installing the driver once after
cluster bringup.

/cc @msau42

**Review by oomichi — COMMENTED** [review]
The implementation seems simple and the unit test coverage looks good for me.
Just one comment.

File: test/e2e/framework/create.go
Comment (pohly) [reply test/e2e/framework/create.go]: @oomichi: I think it makes sense to export it. The load/patch/create cycle intentionally exposes each step via an individual call because then patching can be done differently if desired, and then having the individual building blocks (like patching the namespace, patching container images) for such a patch method available makes sense.

Comment (oomichi) [reply test/e2e/framework/create.go]: @pohly Thanks for your explanation, that makes sense :)

**Review by oomichi — COMMENTED** [review]
/lgtm

File: test/utils/image/manifest.go
Comment (listx) [inline test/utils/image/manifest.go]: Do we already check if registryAndUser is a valid registry/user string? The `countParts == 1` feels a bit too optimistic to me.

File: test/e2e/framework/create.go
Comment (timothysc) [reply test/e2e/framework/create.go]: Move this under https://github.com/kubernetes/kubernetes/tree/master/test/e2e/framework/pod and then lgtm.

**Review by timothysc — CHANGES_REQUESTED** [review]
/assign @timothysc

File: test/utils/image/manifest.go
Comment (johscheuer) [reply test/utils/image/manifest.go]: AFAIK currently there is no check if `registry/user` is actually a valid registry. I think this is also the default behaviour of the docker daemon (without judging if this is a good idea)

File: test/e2e/framework/create.go
Comment (johscheuer) [reply test/e2e/framework/create.go]: @timothysc I moved it under https://github.com/kubernetes/kubernetes/tree/master/test/e2e/framework/pod if you want to take a look again.

File: test/utils/image/manifest.go
Comment (timothysc) [inline test/utils/image/manifest.go]: I don't see why this is necessary. 

/assign @johnSchnake

Comment (johscheuer) [reply test/utils/image/manifest.go]: It seems like this one came in when I rebased it, I will remove these lines

Comment (johscheuer) [reply test/utils/image/manifest.go]: A short grep show that this registry is actually used:

```bash
./test/e2e/testing-manifests/storage-csi/gce-pd/node_ds.yaml:          image: gcr.io/gke-release/csi-driver-registrar:v1.0.1-gke.0
./test/e2e/testing-manifests/storage-csi/gce-pd/node_ds.yaml:          image: gcr.io/gke-release/gcp-compute-persistent-disk-csi-driver:v0.4.0-gke.0
./test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml:          image: gcr.io/gke-release/csi-attacher:v1.0.1-gke.0
./test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml:          image: gcr.io/gke-release/gcp-compute-persistent-disk-csi-driver:v0.4.0-gke.0
```

Comment (johscheuer) [reply test/utils/image/manifest.go]: Since it's not listed in https://github.com/kubernetes/kubernetes/blob/master/test/utils/image/manifest.go#L28 I would remove it here (also I can't see that the files above are used some where). I will open a follow up PR if these files are not used anymore.

Comment (johscheuer) [reply test/utils/image/manifest.go]: and https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/csi.go#L415 uses these files so we need this one

**Review by timothysc — APPROVED** [review]
please squash commits then lgtm 

/approve

**Comment by johscheuer** [comment]
rebased, thanks!

**Review by pohly — APPROVED** [review]
/lgtm

**Pull request merged**
